dxp/javascript_minification.md

```diff
@@ -0,0 +1,86 @@
+    -   `minifier.inline.content.cache.enabled`: set to `true` to make the filter minify `<script>` tags content (by default `true`).
+    -   `minifier.inline.content.cache.skip.javascript`: this defines a comma separated list of tokens to look for in the `<script>` content so that, if any is found, the minified response is not cached. By default, any `<script>` containing `getSessionId` or `encryptedUserId` is not cached.
+    -   `strip.ignore.paths`: list of paths to ignore (by default `/document_library/get_file`).
+    -   `strip.mime.types` in `portal.properties`: defines the mime types of responses that should be processed by the filter (by default `text/html*` and `text/xml*`). Note that you can put a `*` at the end of the config values to catch all mime types starting with the string (not being exactly the string).
+-   Request parameters:
+    -   `ensureContentLength`: set to `true` if you want to make sure the `Content-Length` header is returned from the server.
+    -   `strip`: set to `false` to disable the filter.
+    -   `p_p_lifecycle`: if it is `1` ([PortletRequest.ACTION_PHASE](https://docs.oracle.com/cd/E13155_01/wlp/docs103//javadoc/javax/portlet/PortletRequest.html#ACTION_PHASE)) or `2` ([PortletRequest.RESOURCE_PHASE](https://docs.oracle.com/cd/E13155_01/wlp/docs103//javadoc/javax/portlet/PortletRequest.html#RESOURCE_PHASE)) the filter is disabled.
+
+If all the rules match and a `<script>` is minified, the filter simply invokes [MinifierUtil.minifyJavaScript()](https://github.com/liferay/liferay-portal/blob/master/portal-impl/src/com/liferay/portal/minifier/MinifierUtil.java#L45) to do its duty.
+
+### AggregateFilter
+
+This filter is already described in [Resource Serving in DXP](resource_serving.md#aggregate-filter).
+
+### ComboServlet
+
+This servlet aggregates several `.js` or `.css` files and serves them as a single request to ease caching and reduce connection count. It also caches its responses.
+
+It honors the `work.dir.override.enabled` portal property so if it is set to `true` it diverts requests to development project folders when using fast deployments.
+
+The way to invalidate this servlet's cache is to check if any of the aggregated files has changed. If the portal property `combo.check.timestamp` is set to `false` (and by default it is in production systems) the check is not performed and the cache is never invalidated. Otherwise files are tested for changes every `combo.check.timestamp.interval` milliseconds (where `combo.check.timestamp.interval` is a configuration value of `portal.properties`).
+
+To finish with, each `.js` file aggregated by the servlet has its name tested for the `-min.js` or `.min.js` suffixes and, if they are NOT found, its content is minified using [MinifierUtil.minifyJavaScript()](https://github.com/liferay/liferay-portal/blob/master/portal-impl/src/com/liferay/portal/minifier/MinifierUtil.java#L45).
```

Comment:
"To finish with" → "Finally"

Also another line-specific link here that could be robustified by pointing it at a concrete blob.

Reply:
:thinking: I'm pointing things to `master` because they are fairly stable and, in case of small changes, you see the newest copy and it can enrich the documentation. 

This has the drawback that the link can also be broken, but it is very probable that if the links break because the files have been (re)moved the documentation becomes obsolete, so I decided in the end to use `master` links as the best trade-off...

Reply:
Another option is not pointing to the line but to the file (I used that technique in some other places...)

Reply:
> 🤔 I'm pointing things to `master` because they are fairly stable and, in case of small changes, you see the newest copy and it can enrich the documentation.

This is a bad idea. It risks destroying information, or at least makes it hard to reconstruct:

-   If `master` changes, I can land in the file and see that the line I'm at obviously isn't the one you were talking about.
-   I then have to perform a difficult search to find out what master looked like at the time you published this document, which means looked at the date the line entered the documentation repo, and then finding the corresponding hash on master.
-   Worse still, if we rename or move the files (which has happened multiple times in recent years even for files that don't themselves get touched, as we move massive parts of the directory tree up or down the hierarchy) then finding the document you linked to is even more cumbersome.

Contrast that with the opposite, which I am suggesting:

-   I land on "stale" documentation. I see a URL like https://github.com/liferay/liferay-portal/blob/ab15fb32bd58e4f218c8270be54ba8a835e253b7/portal-impl/src/com/liferay/portal/minifier/MinifierUtil.java#L45 with a hash, so I know I may not looking at current `master`.
-   Based on the header metadata, I decide whether I want to look to see for the current state:

    ![Screenshot 2020-09-11 -154310-87BzyZe9@2x](https://user-images.githubusercontent.com/7074/92932675-8173eb80-f445-11ea-9f12-6f0bc1d7b461.png)
-   If I do, I double-click the SHA1 in the URL bar, type `master`, hit Enter.
-   If stuff has moved, I can hunt around for it.

I previously [managed to convince](https://github.com/liferay/liferay-frontend-guidelines/pull/29#discussion_r293012635) @julien and @jbalsas to do this (although I can't find an example right now, I've seen him giving people this same advice). I probably worded it better back then:

> I suggest hitting "y" on GitHub when you're grabbing links like:
>
> That turns it into an "evergreen" URL that will never break unless we rewrite history (which we won't):
>
> It's all too easy for these links to become dead and broken at the next rename, and that's worse than looking at something slightly stale, I think.

Evergreen links are better, and one of the great things about an immutable version control system is you actually can post links that are immune to bit-rot.

Reply:
The only problem with that reasoning is that some people may not notice that they are on a given commit, instead of master's HEAD and think that code is the latest.

But it is true that it looks like a better scenario, still, so I'll change live links to fixed ones.

The thing is, if we decide to go that way, we should probably use fixed links for everything and don't mix them, to avoid fooling readers.

If, for some reason, we want to use a live link in some place, we should state in that place that the link is live and may change without the documentation being updated.

Does that sounds like a good strategy for the whole repo? :thinking:

Reply:
> Does that sounds like a good strategy for the whole repo? 🤔

I've been trying to paste fixed links for a while now. They do feel safer in the sense that they reflect the state of the code at the time of writing, which makes the article coherent at the very least.

While the code might be stale, so does the article, so this sounds to me like a more robust strategy.

Reply:
My general policy:

- If I am linking to a specific line, _always_ use a fixed link.
- If I am linking to a specific file, use my judgment:
  - If it is possible that the file might be renamed (eg. any source code file), I prefer fixed links.
  - If it is extremely unlikely that a file will be renamed (eg. a `package.json` or a top-level `README.md`), I don't use a fixed link.

I prefer simple rules (not of the form "if A then B else C unless D") because they are easily explained, easily understood, easily enforced etc so it pains me to write that out... you can kind of squint at this one and just reduce it to "favor using fixed links".

![forced-smile](https://user-images.githubusercontent.com/7074/93073092-aacd8b00-f682-11ea-8461-7b4254872bcf.gif)